src/main/kotlin/com/openlattice/data/storage/IndexingMetadataManager.kt

```diff
@@ -37,103 +86,121 @@ class IndexingMetadataManager(private val hds: HikariDataSource) {
+                    val partitionsArray = PostgresArrays
+                            .createIntArray(connection, entitySetPartitions.getValue(entitySetId).partitions)
+                    stmt.setObject(1, entitySetId)
+                    stmt.setArray(2, partitionsArray)
+                    stmt.setArray(3, linkingIdsArray)
+
+                    stmt.addBatch()
+                }
+
+                return stmt.executeBatch().sum()
             }
         }
     }
 
-    fun markAsNeedsToBeLinked(entityDataKeys: Set<EntityDataKey>): Int {
-        return hds.connection.use { connection ->
-            connection.prepareStatement(markAsNeedsToBeLinkedSql(entityDataKeys)).use { stmt ->
-                stmt.executeUpdate()
+    fun markAsNeedsToBeLinked(linkingEntityDataKeys: Set<EntityDataKey>): Int {
+        val linkingEntityKeys = linkingEntityDataKeys
+                .groupBy { it.entitySetId }
+                .mapValues {
+                    it.value.map { it.entityKeyId }
+                }
+        val entitySetPartitions = partitionManager.getEntitySetsPartitionsInfo(linkingEntityKeys.keys)
+
+        hds.connection.use { connection ->
+            connection.prepareStatement(markAsNeedsToBeLinkedSql).use { stmt ->
+                linkingEntityKeys.forEach { (entitySetId, linkingIds) ->
+                    val partitionsArray = PostgresArrays
+                            .createIntArray(connection, entitySetPartitions.getValue(entitySetId).partitions)
+                    val linkingIdsArray = PostgresArrays.createUuidArray(connection, linkingIds)
+                    stmt.setObject(1, entitySetId)
+                    stmt.setArray(2, partitionsArray)
+                    stmt.setArray(3, linkingIdsArray)
+                    stmt.addBatch()
+                }
+
+                return stmt.executeBatch().sum()
             }
         }
     }
 }
 
 /**
  * Arguments of preparable sql in order:
- * 1. entity set id
- * 2. entity key id
- * 3. if linking      -> last_linking_index
- *    if non-linking  -> last_index
+ * 1. last index
+ * 2. entity set id
+ * 3. entity key ids (uuid array)
+ * 4. partition
  */
-fun updateLastIndexSql(linking: Boolean): String {
-    val idColumn = if (linking) LINKING_ID.name else ID.name
-    val filterLinkingIds = if (linking) " AND ${LINKING_ID.name} IS NOT NULL " else ""
-    val entitiesClause = " AND ${ENTITY_SET_ID.name} = ? AND  $idColumn = ? $filterLinkingIds"
-    val withClause = buildWithClause(linking, entitiesClause)
-    // rather use id than linking_id in linking join
-    val joinClause = joinClause(IDS.name, FILTERED_ENTITY_KEY_IDS, entityKeyIdColumnsList)
-    val updateColumn = if (linking) LAST_LINK_INDEX.name else LAST_INDEX.name
+private val updateLastIndexSql = "UPDATE ${IDS.name} SET ${LAST_INDEX.name} = ? " +
+        "WHERE ${VERSION.name} > 0 AND ${ENTITY_SET_ID.name} = ? AND ${ID.name} = ANY(?) AND ${PARTITION.name} = ?"
```

Comment:
I don't think the version > 0 filter is necessary -- could potentially cause weirdness if the entity is deleted after it's indexed and before it's marked as indexed

Reply:
👍